pkg/netutil/loopback.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2018 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package netutil
+
+import (
+	"net"
+	"net/url"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// defaultLoopbackIPs contains all loopback interface addresses.
+// (https://github.com/apple/cups/blob/818bbe7af2c235dc975249ae1231eedd8353de64/scheduler/client.c#L3559-L3562)
+//
+// CVE-2018-5702:
+// - https://bugs.chromium.org/p/project-zero/issues/detail?id=1447#c2
+// - https://github.com/transmission/transmission/pull/468
+var (
+	lmu                sync.RWMutex
+	defaultLoopbackIPs = map[string]struct{}{
+		"localhost":  {}, // localhost is loopback network interface
+		"localhost.": {}, // localhost is loopback network interface
+		"127.0.0.1":  {}, // IPv4
+		"[::1]":      {}, // IPv6
+		"::1":        {}, // IPv6
+	}
+)
+
+func init() {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return
+	}
+	for _, i := range ifaces {
+		addrs, err := i.Addrs()
+		if err != nil {
+			return
+		}
+		for _, addr := range addrs {
+			var ip net.IP
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			if ip.IsLoopback() {
+				lmu.Lock()
+				defaultLoopbackIPs[ip.String()] = struct{}{}
+				lmu.Unlock()
+			}
+		}
+	}
+}
+
+// GetHost removes all schemes and ports from address
+// and only returns the host in the address.
+func GetHost(addr string) (host string) {
```

Comment:
nit: GetHost should probably take an http.Request so it's clearer what string it's acting on.